internal/webhooks/clusterclass.go

```diff
@@ -168,6 +176,147 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
+
+	// Get the old ClusterClassVariables as a map
+	oldVars, _ := getClusterClassVariablesMapWithReverseIndex(old.Spec.Variables)
+
+	// Get the old ClusterClassVariables as a map with an index linking them to their place in the ClusterClass Variable array.
+	// Note: The index is used to improve the error recording below.
+	newVars, clusterClassVariablesIndex := getClusterClassVariablesMapWithReverseIndex(new.Spec.Variables)
+
+	// Retrieve the set of ClusterClass variables that require validation.
+	varsForValidation := getClusterClassVariablesForValidation(oldVars, newVars)
+
+	// Validate the ClusterClassVariables for every ClusterClassVariable on every Cluster using the ClusterClass.
+	for _, cluster := range clusters {
+		for _, clusterVar := range cluster.Spec.Topology.Variables {
+			var varIndex int
+			// Find the varIndex of the variable in the ClusterClass
+			if v, ok := clusterClassVariablesIndex[clusterVar.Name]; ok {
+				varIndex = v
+			}
+
+			// Add Cluster Variable entry in clusterVariableReferences to track where it is in use.
+			if _, ok := clusterVariableReferences[clusterVar.Name]; !ok {
+				clusterVariableReferences[clusterVar.Name] = []string{}
+			}
+			clusterVariableReferences[clusterVar.Name] = append(clusterVariableReferences[clusterVar.Name], cluster.Name)
+
+			// 1) Error if a variable with a schema altered in the update is no longer valid on the Cluster.
+			if alteredVar, ok := varsForValidation.altered[clusterVar.Name]; ok {
+				allErrs = append(allErrs, variables.ValidateClusterVariable(clusterVar.DeepCopy(), alteredVar, path.Index(varIndex),
+					fmt.Sprintf("variable %s in Cluster %s is no longer valid", alteredVar.Name, cluster.Name))...)
+			}
+
+			// 2) Error if a variable removed in the update is still in use in some Clusters.
+			if _, ok := varsForValidation.removed[clusterVar.Name]; ok {
+				allErrs = append(allErrs, field.Invalid(path.Index(varIndex),
+					string(clusterVar.Value.Raw),
+					// TODO: find a way to aggregate this error.
+					fmt.Sprintf("variable %v can not be removed. It is used in Cluster %v", clusterVar.Name, cluster.Name),
+				))
+			}
+
+			// 3) Error if a variable has been added in the update check is no longer valid on the Cluster.
+			// NOTE: This can't occur in normal circumstances as a variable must be defined in a ClusterClass in order to be introduced in
+			// a Cluster. This check may catch errors in cases involving broken Clusters.
+			if addedVar, ok := varsForValidation.added[clusterVar.Name]; ok {
+				allErrs = append(allErrs,
+					variables.ValidateClusterVariable(clusterVar.DeepCopy(), addedVar, path.Index(varIndex),
+						fmt.Sprintf("variable %s in Cluster %s is no longer valid", addedVar.Name, cluster.Name))...)
+			}
+		}
+	}
+
+	// 4) Error if a required variable is not defined in every cluster using the ClusterClass.
+	for varName := range varsForValidation.required {
+		if len(clusterVariableReferences[varName]) != len(clusters) {
```

Comment:
In order to make this consistent with the above, I will loop on all clusters and keep track of the Cluster where the variable is missing (while instead now we are keeping track of the Clusters where the variable is defined)